donkeycar/parts/telemetry.py

```diff
@@ -0,0 +1,111 @@
+    '''
+    Telemetry class collects telemetry from different parts of the system and periodically sends updated to the server.
+    Telemetry reports are timestamped and stored in memory until it is pushed to the server
+    '''
+
+    def __init__(self, cfg, default_inputs = [], default_types = []):
+
+        StreamHandler.__init__(self)
+
+        self.PUBLISH_PERIOD = cfg.TELEMETRY_PUBLISH_PERIOD
+        self._last_publish = time.time()
+        self._telem_q = queue.Queue()
+
+        self._default_inputs = default_inputs
+        self._default_types = default_types
+        self._total_updates = 0
+        self._donkey_name = os.environ.get('DONKEY_NAME', cfg.TELEMETRY_DONKEY_NAME)
+        self._mqtt_broker = os.environ.get('DONKEY_MQTT_BROKER', cfg.TELEMETRY_MQTT_BROKER_HOST) # 'iot.eclipse.org'
+        self._topic = cfg.TELEMETRY_MQTT_TOPIC_TEMPLATE % self._donkey_name
+        self._use_json_format = cfg.TELEMETRY_MQTT_JSON_ENABLE
+        self._mqtt_client = MQTTClient()
+        self._mqtt_client.connect(self._mqtt_broker, cfg.TELEMETRY_MQTT_BROKER_PORT)
+        self._mqtt_client.loop_start()
+        print("Telemetry MQTT server connected.")
+
+    def report(self, metrics):
+        """
+        Basic reporting - gets arbitrary dictionary with values
+        """
+
+        curr_time = int(time.time())
+
+        # Store sample with time rounded to second
+        self._telem_q.put((curr_time, metrics), block=False)
+
+        return curr_time
+
+    def emit(self, record):
+        """
+        FUTURE: Added to support Logging interface (to allow to use Python logging module to log directly to telemetry)
+        """
+        # msg = self.format(record)
+        self.report(record)
+
+    @property
+    def qsize(self):
+        return self._telem_q.qsize()
+
+    def periodic_publish(self):
+
+        curr_time = time.time()
+        if curr_time - self._last_publish > self.PUBLISH_PERIOD and self.qsize > 0:
+
+            # Create packet
+            packet = {}
+            while not self._telem_q.empty():
+                next_item = self._telem_q.get()
+                packet.setdefault(next_item[0], {}).update(next_item[1])
+
+            if self._use_json_format:
+                packet = [{'ts': k, 'values': v} for k, v in packet.items()]
+                payload = json.dumps(packet)
+
+                self._mqtt_client.publish(self._topic, payload)
+                # print('Total updates - %d (payload size=%d)' % (self._total_updates, len(payload)))
+            else:
+                # Publish only the last timestamp
+                last_sample = packet[list(packet)[-1]]
+                for k, v in last_sample.items():
+                    self._mqtt_client.publish('{}/{}'.format(self._topic, k), v)
+                # print('Total updates - %d (values =%d)' % (self._total_updates, len(last_sample)))
+
+            self._last_publish = curr_time
+            self._total_updates += 1
+
+    def run(self, *args):
```

Comment:
Wouldn't it be safer to pass in by `**kwargs` and check that the argument names match `self._default_inputs` and their types match `self._default_types`?

Reply:
Are you sure about it? I implemented it the same way it is done in `TubWriter`. I'd consider it an overkill to run a loop and verify types on such a critical functions when inputs are fully automated...